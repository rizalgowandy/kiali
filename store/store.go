package store

// Store is a generic key value store. Implementations should be safe for concurrent use.
// It is basically just a map with a lock.
type Store[K comparable, V any] interface {
	// Get returns the value associated with the given key. Returns false if key was not found.
	Get(key K) (V, bool)

	// Items returns the store's contents.
	Items() map[K]V

	// Keys returns all the keys in the store.
	Keys() []K

	// Remove removes the given key from the store. If the key does not exist, it does nothing.
	Remove(key K)

	// Replace replaces the contents of the store with the given map.
	Replace(map[K]V)

	// Set associates the given value with the given key. It will overwrite any existing value
	// or create a new entry if the key does not exist.
	Set(key K, value V)

	// Version returns the current version of the store. The version is incremented every time the store is modified.
	// It can be used to detect changes to the store.
	Version() uint
}

// New returns a new store safe for concurrent use.
func New[K comparable, V any]() Store[K, V] {
	return &threadSafeStore[K, V]{data: make(map[K]V)}
}
