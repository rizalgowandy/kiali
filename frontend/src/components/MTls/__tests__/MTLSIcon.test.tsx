import * as React from 'react';
import { mount } from 'enzyme';
import { MTLSIcon, MTLSIconTypes } from '../MTLSIcon';
import { mountToJson } from 'enzyme-to-json';
import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import { Provider } from 'react-redux';
import { store } from 'store/ConfigStore';

const mockIcon = (icon: string) => {
  const component = (
    <Provider store={store}>
      <MTLSIcon
        icon={icon}
        iconClassName="className"
        tooltipText="Overlay Test"
        tooltipPosition={TooltipPosition.right}
      />
    </Provider>
  );
  return mount(component);
};

describe('when Icon is LOCK_FULL', () => {
  it('MTLSIcon renders properly', () => {
    const mount = mockIcon(MTLSIconTypes.LOCK_FULL);

    expect(mountToJson(mount)).toBeDefined();
    expect(mountToJson(mount)).toMatchSnapshot();

    const tooltip = mount.find(Tooltip);
    expect(tooltip.exists()).toBeTruthy();
    expect(tooltip.props().position).toEqual('right');
    expect(tooltip.props().content).toEqual('Overlay Test');

    const img = tooltip.find('img');
    expect(img.exists()).toBeTruthy();
    expect(img.props().className).toEqual('className');
    expect(img.props().src).toEqual('mtls-status-full-dark.svg');
  });
});

describe('when Icon is LOCK_HOLLOW', () => {
  it('MTLSIcon renders properly', () => {
    const mount = mockIcon(MTLSIconTypes.LOCK_HOLLOW);
    const img = mount.find('img');
    expect(img.exists()).toBeTruthy();
    expect(img.props().className).toEqual('className');
    expect(img.props().src).toEqual('mtls-status-partial-dark.svg');
  });
});
