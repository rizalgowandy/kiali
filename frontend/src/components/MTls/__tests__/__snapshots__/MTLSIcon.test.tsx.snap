// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when Icon is LOCK_FULL MTLSIcon renders properly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MTLSIcon
    icon="LOCK_FULL"
    iconClassName="className"
    tooltipPosition="right"
    tooltipText="Overlay Test"
  >
    <Tooltip
      aria-label="mTLS status"
      content="Overlay Test"
      enableFlip={true}
      position="right"
    >
      <Popper
        animationDuration={300}
        appendTo={[Function]}
        distance={15}
        enableFlip={true}
        entryDelay={300}
        exitDelay={300}
        flipBehavior={
          Array [
            "top",
            "right",
            "bottom",
            "left",
            "top",
            "right",
            "bottom",
          ]
        }
        isVisible={false}
        minWidth="revert"
        onBlur={[Function]}
        onDocumentClick={false}
        onDocumentKeyDown={[Function]}
        onFocus={[Function]}
        onHidden={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onPopperMouseEnter={[Function]}
        onPopperMouseLeave={[Function]}
        onTriggerEnter={[Function]}
        placement="right"
        popper={
          <div
            aria-label="mTLS status"
            aria-live="off"
            className="pf-v5-c-tooltip"
            id="pf-tooltip-2"
            role="tooltip"
            style={
              Object {
                "maxWidth": null,
              }
            }
          >
            <TooltipArrow />
            <TooltipContent
              isLeftAligned={false}
            >
              Overlay Test
            </TooltipContent>
          </div>
        }
        popperRef={
          Object {
            "current": null,
          }
        }
        positionModifiers={
          Object {
            "bottom": "pf-m-bottom",
            "bottom-end": "pf-m-bottom-right",
            "bottom-start": "pf-m-bottom-left",
            "left": "pf-m-left",
            "left-end": "pf-m-left-bottom",
            "left-start": "pf-m-left-top",
            "right": "pf-m-right",
            "right-end": "pf-m-right-bottom",
            "right-start": "pf-m-right-top",
            "top": "pf-m-top",
            "top-end": "pf-m-top-right",
            "top-start": "pf-m-top-left",
          }
        }
        trigger={
          <img
            alt="right"
            className="className"
            src="mtls-status-full-dark.svg"
          />
        }
        zIndex={9999}
      >
        <div
          style={
            Object {
              "display": "contents",
            }
          }
        >
          <img
            alt="right"
            className="className"
            key="mtls-status-full-dark.svg"
            src="mtls-status-full-dark.svg"
          />
        </div>
      </Popper>
    </Tooltip>
  </MTLSIcon>
</Provider>
`;
